ocaml/util/tar_helpers.ml

```diff
@@ -0,0 +1,80 @@
+module Identity = struct
+  type 'a t = 'a
+  let return x = x
+  let ( >>= ) m f = f m
+end
+
+module Writer = struct
+  type out_channel = Unix.file_descr
+  type 'a t = 'a Identity.t
+
+  let really_write fd buffer =
+    let s = Cstruct.to_string buffer |> Bytes.of_string in
+    if Bytes.length s > 0
+    then ignore (Unix.write fd s 0 (Bytes.length s))
+end
+module HW = Tar.HeaderWriter(Identity)(Writer)
+
+let rec with_restart op fd buf off len =
+  try
+    op fd buf off len
+  with Unix.Unix_error (Unix.EINTR,_,_) ->
+    (with_restart [@tailcall]) op fd buf off len
+  
+
+let rec _really_input fd buf off = function
+  | 0 -> ()
+  | len ->
+    let m = Unix.read fd buf off len in
+    if m = 0 then (
+      raise End_of_file
+    );
+    _really_input fd buf (off+m) (len-m)
+
+let really_input = with_restart _really_input
+
+let really_read ifd buffer =
+  let s = Bytes.create (Cstruct.len buffer) in
+  really_input ifd s 0 (Cstruct.len buffer);
+  Cstruct.blit_from_bytes s 0 buffer 0 (Cstruct.len buffer)
+
+let skip ifd n =
+  let buffer = Cstruct.create 4096 in
+  let rec loop n =
+    if n <= 0 then ()
+    else (
+      let amount = min n (Cstruct.len buffer) in
+      really_read ifd (Cstruct.sub buffer 0 amount);
+      loop (n - amount)
+    )
+  in
+  loop n
+
+let copy_n ifd ofd n =
+  let buffer = Bytes.create 16384 in
+  let rec loop remaining =
+  if remaining = 0L then ()
+  else (
+    let this = Int64.(to_int (min (of_int (Bytes.length buffer)) remaining)) in
+    let n = input (Unix.in_channel_of_descr ifd) buffer 0 this in
+    if n = 0 then 
+      raise End_of_file;
+    (
+      try
+        output (Unix.out_channel_of_descr ofd) buffer 0 n
```

Comment:
Similarly this creates a new output channel every time we go around the loop. Who flushes it? Who closes it?
I think that the channels should be created once outside of the loop, and then at the end the out channel should be flushed, but even then since you're doing your own buffering here anyway could just use the Unix.read/Unix.write functions directly.